lib/src/modules/task.cc

```diff
@@ -128,11 +136,171 @@ static TaskCommand getTaskCommand(const fs::path &task_executable)
+    url += "?";
+    for (auto& key : params.keys()) {
+      updateUrl(url, curl, key);
+      url += "=";
+      updateUrl(url, curl, params.get<std::string>(key));
+      url += "&";
+    }
+    return url;
+}
+
+// This method downloads the file at the specified url into the provided path
+// If something goes wrong during the download, a PXP-error should be returned.
+static void downloadTaskFile(const std::vector<std::string>& master_uris,
+                             const fs::path& path,
+                             const lth_jc::JsonContainer& uri) {
+    auto curl = curl_easy_init();
+    if (!curl) {
+      throw Module::ProcessingError(lth_loc::format("Failed to initiate the task file download process"));
+    }
+    char err_buff[CURL_ERROR_SIZE];
+    auto fp = fopen(path.string().c_str(), "wb");
+    auto url = master_uris[0] + createUrlEndpoint(curl, uri);
+    curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
+    curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, writeData);
+    curl_easy_setopt(curl, CURLOPT_WRITEDATA, fp);
+    // We can get an error message this way if something goes wrong in the download
+    curl_easy_setopt(curl, CURLOPT_ERRORBUFFER, err_buff);
+    // CONNECTTIMEOUT specifies how long curl should wait for a connection to be established.
+    // Default is 300 seconds.
+    curl_easy_setopt(curl, CURLOPT_CONNECTTIMEOUT, 300);
+    auto res = curl_easy_perform(curl);
+    curl_easy_cleanup(curl);
+    fclose(fp);
+    if (res != CURLE_OK) {
+        fs::remove(path);
+        throw Module::ProcessingError(err_buff);
+    }
+}
+
+// This method does the following. If the file matching the "filename" field of the
+// file_obj JSON does not exist OR if its hash does not match the sha value in the
+// "sha256" field of file_obj, then:
+//    (1) The file is downloaded using libcurl from the first entry in master-uris
+//        to a temporary directory. If this download fails, a PXP error is thrown.
+//
+//    (2) If the downloaded file's sha does not match the provided sha, then a PXP
+//        error is returned.
+//
+//    (3) If (1) and (2) both succeed, then the downloaded file is atomically
+//        renamed to cache_dir/<filename>
+static fs::path updateTaskFile(const std::vector<std::string>& master_uris,
+                               const fs::path& cache_dir,
+                               const lth_jc::JsonContainer& file) {
+    auto filename = file.get<std::string>("filename");
+    auto sha256 = file.get<std::string>("sha256");
+    auto filepath = cache_dir / filename;
+    // TODO: Also check permissions? Make sure it's readable/writable/executable by
+    // PXP-agent's owner/group?
+    if (fs::exists(filepath) && sha256 == calculateSha256(filepath.string())) {
+      return filepath;
+    }
+
+    // TODO: Maybe place somewhere else? A task module without any master uris
+    // would need to have all its files hardcoded in its respective <sha> directory.
+    if (master_uris.empty()) {
```

Comment:
This seems reasonable to do only if we actually need to download a file. That appears to be what you've done.